dashboard/modules/metric_history.py

```diff
@@ -0,0 +1,207 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from math import pi
+
+from bokeh.models import CategoricalColorMapper, ColumnDataSource, CustomJS, HoverTool, TapTool, Whisker
+from bokeh.plotting import figure
+
+import javascript_utils
+import utils
+
+import numpy as np
+
+
+# For a given test_name, find the history of metrics for that test.
+QUERY = """
+SELECT
+  metrics.test_name,
+  metrics.metric_name,
+  SAFE_CAST(EXTRACT(date from metrics.timestamp) AS STRING) AS run_date,
+  metrics.metric_value,
+  metrics.metric_lower_bound,
+  metrics.metric_upper_bound,
+  job.job_status,
+  job.stackdriver_logs_link AS logs_link
+FROM (
+  SELECT
+    x.test_name,
+    x.metric_name,
+    x.timestamp,
+    x.metric_value,
+    x.metric_lower_bound,
+    x.metric_upper_bound,
+    x.uuid
+  FROM (
+    SELECT
+      test_name,
+      metric_name,
+      SAFE_CAST(EXTRACT(date from timestamp) AS STRING) as run_date,
+      min(timestamp) as min_timestamp
+    FROM
+      `xl-ml-test.metrics_handler_dataset.metric_history`
+    WHERE
+      test_name = '%(test_name)s' AND
+      (metric_name NOT LIKE '%%__Percentile_%%' OR metric_name LIKE '%%__Percentile_99%%')
+    GROUP BY
+      test_name, metric_name, run_date
+  ) AS y
+  INNER JOIN `xl-ml-test.metrics_handler_dataset.metric_history` AS x
+  ON
+    y.test_name = x.test_name AND
+    y.metric_name = x.metric_name AND
+    y.min_timestamp = x.timestamp
+) AS metrics
+INNER JOIN `xl-ml-test.metrics_handler_dataset.job_history` AS job
+ON
+  metrics.uuid = job.uuid
+ORDER BY
+  run_date DESC
+"""
+
+
+def fetch_data(test_name):
+  dataframe = utils.run_query(QUERY % {'test_name': test_name},
```

Comment:
Can you use BQ parameterized queries here (and for other queries) to avoid SQL injection concerns?

https://googleapis.dev/python/bigquery/latest/usage/queries.html#run-a-query-using-a-named-query-parameter
https://googleapis.dev/python/bigquery/latest/usage/queries.html#run-a-query-using-a-named-query-parameter

Reply:
Switched to `pandas-gbq` parameterized queries here and in `main_heatmap.py`